pkg/storage/stores/shipper/downloads/table.go

```diff
@@ -0,0 +1,400 @@
+package downloads
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/cortexproject/cortex/pkg/chunk"
+	"github.com/cortexproject/cortex/pkg/chunk/local"
+	chunk_util "github.com/cortexproject/cortex/pkg/chunk/util"
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/go-kit/kit/log/level"
+	"go.etcd.io/bbolt"
+)
+
+// timeout for downloading initial files for a table to avoid leaking resources by allowing it to take all the time.
+const downloadTimeout = 5 * time.Minute
+
+type BoltDBIndexClient interface {
+	QueryDB(ctx context.Context, db *bbolt.DB, query chunk.IndexQuery, callback func(chunk.IndexQuery, chunk.ReadBatch) (shouldContinue bool)) error
```

Comment:
I think you can use `chunk_util.Callback` here.

Reply:
I had tried it. The problem is the function in boltdb index client in Cortex defines function using `callback func(chunk.IndexQuery, chunk.ReadBatch) (shouldContinue bool)` instead of `chunk_util.Callback`. If I change it here without changing in Cortex, interface is not satisfied anymore. We will have to change it in Cortex first.

Reply:
ho then just define it here.

Reply:
Sorry, didn't get you. Don't we have to change it in Cortex first?